### PR TITLE
KAFKA-9425: InFlightRequests Class Uses Thread-Safe Counter Non-Threa…

### PR DESCRIPTION
**Pull request by belugabehr**
…d-Safe Collection

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/InFlightRequests.java [hunk 1]
```diff
@@ -32,8 +31,7 @@
 
     private final int maxInFlightRequestsPerConnection;
     private final Map<String, Deque<NetworkClient.InFlightRequest>> requests = new HashMap<>();
-    /** Thread safe total number of in flight requests. */
-    private final AtomicInteger inFlightRequestCount = new AtomicInteger(0);
+    private int inFlightRequestCount = 0;
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/InFlightRequests.java]: This is intentionally thread-safe as per the documentation, so it's not safe to change it.

File: clients/src/main/java/org/apache/kafka/clients/InFlightRequests.java [hunk 2]
```diff
@@ -117,24 +113,24 @@ public int count(String node) {
      * Return true if there is no in-flight request directed at the given node and false otherwise
      */
     public boolean isEmpty(String node) {
-        Deque<NetworkClient.InFlightRequest> queue = requests.get(node);
-        return queue == null || queue.isEmpty();
+        return count(node) == 0;
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/InFlightRequests.java]: Generally, replacing an `isEmpty` check with `count` is not a good idea as you can easily regress in performance. In this case, the deque implementation may have a cheap `size()`, but it's brittle and not something I think we should do.

**Review by ijuma — COMMENTED** [review]
Thanks for the PR. I am not sure why we would want to introduce a data race. The field is used by multiple threads and hence we use an atomic integer for it.

File: clients/src/main/java/org/apache/kafka/clients/InFlightRequests.java (hunk 1 above)
Comment (belugabehr) [reply clients/src/main/java/org/apache/kafka/clients/InFlightRequests.java]: Yes.  I saw the comment.  I'm not sure what the rational is. though. Multiple threads can access the same native int OK.  Can make it volatile if that is preferable.  Is it just that reading can take place multi-threaded?  Clearly the increment/decrement are not multi-threaded operations

File: clients/src/main/java/org/apache/kafka/clients/InFlightRequests.java (hunk 2 above)
Comment (belugabehr) [reply clients/src/main/java/org/apache/kafka/clients/InFlightRequests.java]: OK.

File: clients/src/main/java/org/apache/kafka/clients/InFlightRequests.java (hunk 1 above)
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/clients/InFlightRequests.java]: Yes, reading can be done from multiple threads. `volatile` would probably be enough for this use case.

Comment (belugabehr) [reply clients/src/main/java/org/apache/kafka/clients/InFlightRequests.java]: OK.  Changed to volatile.  This seems optimal to me in terms of performance and clarity.

File: clients/src/main/java/org/apache/kafka/clients/InFlightRequests.java [hunk 3]
```diff
@@ -44,22 +42,18 @@ public InFlightRequests(int maxInFlightRequestsPerConnection) {
...
 
     /**
      * Get the request queue for the given node
      */
     private Deque<NetworkClient.InFlightRequest> requestQueue(String node) {
         Deque<NetworkClient.InFlightRequest> reqs = requests.get(node);
-        if (reqs == null || reqs.isEmpty())
+        if (reqs == null || reqs.isEmpty()) {
             throw new IllegalStateException("There are no in-flight requests for node " + node);
+        }
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/InFlightRequests.java]: Why are we making these unrelated formatting changes?

Comment (belugabehr) [reply clients/src/main/java/org/apache/kafka/clients/InFlightRequests.java]: Well, since I'm touching the file anyway, figured we could do these trivial changes.  Different projects have different tolerances for these types of change. I can revert if you'd like.

Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/clients/InFlightRequests.java]: Kafka doesn't mandate braces in `if` statements.

**Comment by ijuma** [comment]
ok to test

**Comment by belugabehr** [comment]
@ijuma Now I see why the original author went for an `AtomicInteger`.  FindBugs does not like this paradigm.  Hmmmm.  Thoughts?  Just forget this exercise in the pedantic?

```
 	Increment of volatile field org.apache.kafka.clients.InFlightRequests.inFlightRequestCount in org.apache.kafka.clients.InFlightRequests.add(NetworkClient$InFlightRequest)
```

**Comment by belugabehr** [comment]
retest this please

**Comment by belugabehr** [comment]
Inspired by some work I saw in OpenJDK, I figured out what FindBugs wants to see.  I added some documentation for future dev's to utilize.

**Comment by ijuma** [comment]
> @ijuma Now I see why the original author went for an `AtomicInteger`. FindBugs does not like this paradigm. Hmmmm. Thoughts? Just forget this exercise in the pedantic?

It seems like this is not worth it as you said.

**Comment by belugabehr** [comment]
@ijuma OK, well I added some JavaDoc explaining why this looks funny (non thread-safe collection and thread-safe counter).  And did a small amount of cleanup.  Please take a look.

File: clients/src/main/java/org/apache/kafka/clients/InFlightRequests.java [hunk 4]
```diff
@@ -26,13 +26,19 @@
 import java.util.concurrent.atomic.AtomicInteger;
 
 /**
- * The set of requests which have been sent or are being sent but haven't yet received a response
+ * The set of requests which have been sent, or are being sent, but have not yet
+ * received a response. This class is not thread-safe as it pertains to
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/InFlightRequests.java]: Let's use a new paragraph for the thread-safety description. Also, what do we mean by `relatively accurate`? I think that's a bit ambiguous. Maybe call it `eventually consistent` or something like that?

File: clients/src/main/java/org/apache/kafka/clients/InFlightRequests.java [hunk 5]
```diff
@@ -26,13 +26,19 @@
...
+ * multi-threaded way.
  */
 final class InFlightRequests {
 
     private final int maxInFlightRequestsPerConnection;
     private final Map<String, Deque<NetworkClient.InFlightRequest>> requests = new HashMap<>();
-    /** Thread safe total number of in flight requests. */
+
+    // Will only be modified by a single thread, but may be accessed by many
+    // threads that wish to read the count
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/InFlightRequests.java]: I think we can be more concise and simply say `// Single writer thread, multiple reader threads`

File: clients/src/main/java/org/apache/kafka/clients/InFlightRequests.java (hunk 3 above)
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/clients/InFlightRequests.java]: Let's remove the brace changes please.

File: clients/src/main/java/org/apache/kafka/clients/InFlightRequests.java [hunk 6]
```diff
@@ -151,16 +154,17 @@ public boolean isEmpty() {
             return Collections.emptyList();
         } else {
             final Deque<NetworkClient.InFlightRequest> clearedRequests = requests.remove(node);
-            inFlightRequestCount.getAndAdd(-clearedRequests.size());
+            inFlightRequestCount.addAndGet(-clearedRequests.size());
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/InFlightRequests.java]: What is the reason for this change?

**Review by ijuma — COMMENTED** [review]
Thanks, a few more comments.

**Comment by belugabehr** [comment]
@ijuma Really appreciate you working with me on the little things.  In my experience, project leaders that care about the little things are sure to also care about the big things. :)

**Comment by belugabehr** [comment]
@ijuma  Any more feedback on this?
